tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php

```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
 
         self::assertNotContains('<div', $subject->render());
     }
+
+    /**
+     * @test
+     */
+    public function removeRedundantClassesProvidesFluentInterface()
+    {
+        $subject = HtmlPruner::fromHtml('<html></html>');
+
+        $result = $subject->removeRedundantClasses([]);
+
+        self::assertSame($subject, $result);
+    }
+
+    /**
+     * @return string[][][]
+     */
+    public function classesToKeepDataProvider()
+    {
+        return [
+            'no classes to keep' => [[]],
+            '1 class to keep' => [['foo']],
+        ];
+    }
+
+    /**
+     * @test
+     *
+     * @param string[] $classesToKeep
+     *
+     * @dataProvider classesToKeepDataProvider
+     */
+    public function removeRedundantClassesPreservesHtmlWithoutClasses(array $classesToKeep)
+    {
+        $html = '<p style="color: green;">hello</p>';
+        $subject = HtmlPruner::fromHtml('<html>' . $html . '</html>');
+
+        $subject->removeRedundantClasses($classesToKeep);
+
+        self::assertContains($html, $subject->render());
+    }
+
+    /**
+     * @return (string|string[])[][]
```

Comment:
I propose making this more redundant, but easier to read:
`string[][]|string[][][]`

Reply:
That wouldn't be correct - it would mean that either all elements of the 2nd level array are strings, or they're all arrays of strings.  Whereas in fact, some of them are strings and some of them are arrays of strings.  The only viable alternative I can think of would be `mixed[][]`...